### PR TITLE
feat(lsp): allow function rename

### PR DESCRIPTION
**Pull request by kobyhallx**
# Description

## Problem\*

Resolves 

Rename function Symbol with LSP #4192

## Summary\*

Allows to rename function symbol.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs [hunk 1]
```diff
@@ -1463,10 +1465,30 @@ impl<'a> Resolver<'a> {
...
+                    if hir_ident.id != DefinitionId::dummy_id() {
+                        if let DefinitionKind::Function(func_def_id) =
+                            self.interner.definition(hir_ident.id).kind
+                        {
+                            if let Some(func_meta) = self.interner.func_meta.get(&func_def_id) {
+                                self.interner.add_reference(
+                                    (DependencyId::Function(func_def_id), func_meta.location),
+                                    (
+                                        DependencyId::CallExpression(func),
+                                        Location::new(call_expr.func.span, self.file),
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/resolution/resolver.rs:1485]: Instead of adding references during a call, it will be simpler to add them when the function variable is referenced in the `ExpressionKind::Variable` case

Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs]: ty e685934: chore: move collection point

File: compiler/noirc_frontend/src/locations.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+            DependencyId::FunctionCall => {
+                let mut edit_locations: Vec<Location> = Vec::new();
+                if let Some(referenced_node_index) = self
+                    .graph_references
+                    .neighbors_directed(node_index, petgraph::Direction::Incoming)
+                    .next()
+                {
+                    edit_locations.extend(self.get_edit_locations(referenced_node_index));
+                }
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply compiler/noirc_frontend/src/locations.rs:92]: We could also `map` and `unwrap_or_default` instead.

**Review by TomAFrench — COMMENTED** [review]
some nits

File: compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs [hunk 3]
```diff
@@ -381,6 +388,24 @@ impl DefCollector {
     }
 }
 
+fn add_import_refrence(
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs]: ab455b1: chore: review nits

File: compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs [hunk 4]
```diff
@@ -381,6 +388,24 @@ impl DefCollector {
...
+    ns: crate::macros_api::ModuleDefId,
+    name: &Ident,
+    interner: &mut NodeInterner,
+    file_id: FileId,
+) {
+    if name.span().start() == 0 && name.span().end() == 0 {
+        // We ignore empty spans at 0 lcoation, this must be Stdlib
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs]: ab455b1: chore: review nits

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs [hunk 5]
```diff
@@ -1417,7 +1417,9 @@ impl<'a> Resolver<'a> {
                     // Otherwise, then it is referring to an Identifier
                     // This lookup allows support of such statements: let x = foo::bar::SOME_GLOBAL + 10;
                     // If the expression is a singular indent, we search the resolver's current scope as normal.
-                    let (hir_ident, var_scope_index) = self.get_ident_from_path(path);
+                    let (hir_ident, var_scope_index) = self.get_ident_from_path(path.clone());
+
+                    // tracing::debug!("Resolved variable: {:?}", hir_ident);
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs]: ab455b1: chore: review nits

File: compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs [hunk 6]
```diff
@@ -381,6 +388,24 @@ impl DefCollector {
     }
 }
 
+fn add_import_refrence(
+    ns: crate::macros_api::ModuleDefId,
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs]: ab455b1: chore: review nits

File: compiler/noirc_frontend/src/locations.rs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+    pub fn check_rename_possible(&self, location: Location) -> bool {
+        self.location_store.get_node_from_location(location).is_some()
+    }
+
+    pub fn find_rename_symbols_at(&self, location: Location) -> Option<Vec<Location>> {
+        let node_index = self.location_store.get_node_from_location(location)?;
+
+        // let mut edit_locations: Vec<Location> = Vec::new();
+
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/locations.rs]: ab455b1: chore: review nits

File: compiler/noirc_frontend/src/locations.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/locations.rs]: ab455b1: chore: review nits

File: compiler/noirc_frontend/src/locations.rs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                self.get_edit_locations(referenced_node_index)
+            }
+        };
+        Some(found_locations)
+    }
+
+    fn get_edit_locations(&self, referenced_node_index: PetGraphIndex) -> Vec<Location> {
+        let mut edit_locations: Vec<Location> = Vec::new();
+        let (_referenced_id, referencing_location) = self.graph_references[referenced_node_index];
+        edit_locations.push(referencing_location);
```
Comment (jfecher) [inline compiler/noirc_frontend/src/locations.rs:97]:
```suggestion
        let (_referenced_id, referencing_location) = self.graph_references[referenced_node_index];
        let mut edit_locations = vec![referencing_location];
```

File: compiler/noirc_frontend/src/node_interner.rs [hunk 9]
```diff
@@ -167,6 +168,13 @@ pub struct NodeInterner {
 
     /// Stores the [Location] of a [Type] reference
     pub(crate) type_ref_locations: Vec<(Type, Location)>,
+
+    /// Store the location of the references in the graph
+    pub(crate) graph_references: DiGraph<(DependencyId, Location), ()>,
+    /// Tracks the index of the references in the graph
+    pub(crate) graph_references_indices: HashMap<DependencyId, PetGraphIndex>,
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/node_interner.rs]: 4421a2d: chore: name change

File: compiler/noirc_frontend/src/node_interner.rs [hunk 10]
```diff
@@ -167,6 +168,13 @@ pub struct NodeInterner {
...
+    /// Store the location of the references in the graph
+    pub(crate) location_store: LocationStore,
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/node_interner.rs]: d564ac8: chore: name change
Thought LocationIndices is most universal? can we store more than references?

File: compiler/noirc_frontend/src/node_interner.rs [hunk 11]
```diff
@@ -1611,6 +1628,7 @@ impl NodeInterner {
             DependencyId::Global(id) => {
                 Cow::Borrowed(self.get_global(id).ident.0.contents.as_ref())
             }
+            _ => Cow::Borrowed(""),
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/node_interner.rs:1631]: aren't we multiplying LOC for no real benefit?

[the remark this answers is not in the document]
Comment (TomAFrench) [reply compiler/noirc_frontend/src/node_interner.rs:1631]: We've had situations in the past where e.g. new blackbox functions aren't handled properly because we used a catch-all in a match statement which masked when we should be adding handling for this new enum variant. This allows us to catch these errors more easily.

In this case in particular this catch-all is only handling a single enum variant so there's not even a change to LOC.

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs [hunk 12]
```diff
@@ -1436,6 +1436,13 @@ impl<'a> Resolver<'a> {
                                         self.interner.function_visibility(id),
                                     );
                                 }
+
+                                if let Some(func_meta) = self.interner.func_meta.get(&id) {
+                                    self.interner.add_reference(
+                                        (DependencyId::Function(id), func_meta.location),
+                                        (DependencyId::FunctionCall, hir_ident.location),
+                                    );
+                                }
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs:1445]: Doing that will produce more code after walking the graph and collecting relations. I was hoping to avoid that by having simple resolution mechanism.
Perhaps graph describing Node types and edges describing type of relationship would suit us better?

**Review by jfecher — APPROVED** [review]
:+1:

File: tooling/lsp/src/requests/rename.rs [hunk 13]
```diff
@@ -160,3 +160,144 @@ fn on_rename_inner(
...
+                    start: Position { line: 70, character: 3 },
+                    end: Position { line: 70, character: 17 },
+                },
+                Range {
+                    start: Position { line: 79, character: 4 },
+                    end: Position { line: 79, character: 18 },
+                },
+                Range {
+                    start: Position { line: 94, character: 4 },
+                    end: Position { line: 94, character: 18 },
```
Comment (TomAFrench) [inline tooling/lsp/src/requests/rename.rs:298]: It would be nice if we pulled these characters from the file and checked that these ranges contained the old name which we want to rename.

Comment (asterite) [reply tooling/lsp/src/requests/rename.rs:298]: Sounds good, I'll do it.

By the way, I got it to work with the elaborator 🎉 One part of the elaborator code seems to have been copied from the resolver, so I applied the same changes there and it worked 😄 

I think there's nothing to do on the `vscode-noir` side: once the rename thing is registered, "Rename symbol" shows up on the editor. I'll send a small video once I solve this particular issue.

Comment (asterite) [reply tooling/lsp/src/requests/rename.rs:298]: Oh, I saw some tests rely on the `7_functions` program... so I did the same thing. But it could be a bit brittle: if for some reason we change that example program, the lsp server might start to fail. But I don't know if existing test programs change at all. Do you think I should be using a separate program for these tests?

**Comment by asterite** [comment]
Here's the rename in action:

https://github.com/noir-lang/noir/assets/209371/184d9b96-1ddc-4d70-b35c-af583faa2004

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs [hunk 14]
```diff
@@ -1512,19 +1512,27 @@ impl<'a> Resolver<'a> {
                     // Otherwise, then it is referring to an Identifier
                     // This lookup allows support of such statements: let x = foo::bar::SOME_GLOBAL + 10;
                     // If the expression is a singular indent, we search the resolver's current scope as normal.
-                    let (hir_ident, var_scope_index) = self.get_ident_from_path(path);
+                    let (hir_ident, var_scope_index) = self.get_ident_from_path(path.clone());
```
Comment (asterite) [inline compiler/noirc_frontend/src/hir/resolution/resolver.rs:1515]: Note: all of the changes in this file could be reverted, given that this code is now unused (the lsp code uses the elaborator). Should we undo this change?

Comment (jfecher) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs:1515]: If it's not needed for this PR then I'd say yes

Comment (jfecher) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs:1515]: Since it is already done, might as well keep it

Comment (asterite) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs:1515]: Oops, sorry, I just see this. I ended up removing it because it's unused and untested.

File: compiler/noirc_frontend/src/node_interner.rs [hunk 15]
```diff
@@ -182,6 +183,15 @@ pub struct NodeInterner {
...
     quoted_types: noirc_arena::Arena<Type>,
+
+    /// Store the location of the references in the graph
+    pub(crate) reference_graph: DiGraph<DependencyId, ()>,
+
+    /// Tracks the index of the references in the graph
+    pub(crate) reference_graph_indices: HashMap<DependencyId, PetGraphIndex>,
+
+    /// Store the location of the references in the graph
+    pub(crate) location_indices: LocationIndices,
```
Comment (asterite) [inline compiler/noirc_frontend/src/node_interner.rs:194]: 💭 One thing I'm thinking is that these locations and references are now tracked in all compilations, even thought they are only needed by the LSP code. Ideally there's a boolean controlling not to track this, but the interner is created deep in the code so getting a boolean from `check_crate` up to that point is kind of tricky. It could be done in a separate PR, though, and only if really needed (if we notice this change is making compilation times slow).

Comment (TomAFrench) [reply compiler/noirc_frontend/src/node_interner.rs:194]: Yes, this is something I'm a little concerned about in the long-term. Having a clean separation between things which are necessary for the compiler to function and "optional" features would be really good.

**Comment by TomAFrench** [comment]
> Here's the rename in action:

If I try this compile using this branch, restart the LSP and then repeat these steps I get an error tooltop saying "this element cannot be renamed"

**Comment by TomAFrench** [comment]
Ah, highlighting the relevant name and then rightclicking that results in the above error. If I just rightclick the name without highlighting then it works.

**Comment by TomAFrench** [comment]
Let's say I have the program below
```rust
fn main()  -> Field {
    foo::bar()
}

mod foo {
    fn bar() -> Field {
        1
    }
}

```

I now want to rename the function `bar` to `baz`. I then click on either the function definition or the call and rename it. I would expect to then get the program

```rust 
fn main()  -> Field {
    foo::baz()
}

mod foo {
    pub fn baz() -> Field {
        1
    }
}
```


Instead I get the program below

```rust 
fn main()  -> Field {
    baz()
}

mod foo {
    pub fn baz() -> Field {
        1
    }
}
```

We've removed the path information in the call to `foo::bar` and so this now doesn't compile. Ideally we'd just rename the last segment of the path rather than replacing the whole thing.

**Comment by TomAFrench** [comment]
We do correctly update import statements so the above works as expected so it's just qualified paths which are the issue.

```rust
use foo::bar;

fn main() -> pub Field {
    bar()
}

mod foo {
    pub fn bar() -> Field {
        1
    }
}
```

**Comment by TomAFrench** [comment]
@asterite I've pushed a commit which separates checking that the hardcoded ranges correspond to what we expect from testing the lsp response. We can then just iterate through all of the instances of `another_function` and check that we can initiate a renaming of all the other instances.

**Comment by TomAFrench** [comment]
We could likely remove the other two tests now as we're checking that we can rename a symbol from any instance of it.

File: compiler/noirc_frontend/src/elaborator/patterns.rs [hunk 16]
```diff
@@ -418,10 +418,14 @@ impl<'context> Elaborator<'context> {
...
             if hir_ident.id != DefinitionId::dummy_id() {
                 match self.interner.definition(hir_ident.id).kind {
-                    DefinitionKind::Function(id) => {
+                    DefinitionKind::Function(func_id) => {
                         if let Some(current_item) = self.current_item {
-                            self.interner.add_function_dependency(current_item, id);
+                            self.interner.add_function_dependency(current_item, func_id);
                         }
+
+                        let variable = DependencyId::Variable(hir_ident.location);
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/elaborator/patterns.rs:426]: To make things more concrete, the issue is that this location contains the span for the full path rather than just the relevant segment.

File: compiler/noirc_frontend/src/locations.rs [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    pub fn check_rename_possible(&self, location: Location) -> bool {
+        self.location_indices.get_node_from_location(location).is_some()
+    }
+
+    pub fn find_rename_symbols_at(&self, location: Location) -> Option<Vec<Location>> {
+        let node_index = self.location_indices.get_node_from_location(location)?;
+
+        let reference_node = self.reference_graph[node_index];
+        let found_locations: Vec<Location> = match reference_node {
+            DependencyId::Alias(_) | DependencyId::Struct(_) | DependencyId::Global(_) => todo!(),
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply compiler/noirc_frontend/src/locations.rs:95]: Added one for structs https://github.com/noir-lang/noir/issues/5352

[the remark this answers is not in the document]
Comment (asterite) [reply compiler/noirc_frontend/src/locations.rs:95]: I added one for type aliases ( https://github.com/noir-lang/noir/issues/5358 ) and one for global variables ( https://github.com/noir-lang/noir/issues/5359 ).

[the remark this answers is not in the document]
Comment (TomAFrench) [reply compiler/noirc_frontend/src/locations.rs:95]: Thanks!

File: compiler/noirc_frontend/src/node_interner.rs (hunk 15 above)
Comment (TomAFrench) [reply compiler/noirc_frontend/src/node_interner.rs:194]: Not required for this PR however to be clear.

File: compiler/noirc_frontend/src/elaborator/patterns.rs (hunk 16 above)
Comment (asterite) [reply compiler/noirc_frontend/src/elaborator/patterns.rs:426]: Thank you for pointing the exact location that needed a change!

I just pushed a commit that fixes this. I also reorganized the tests a bit (I hope it's fine) because I needed to write a new test that was exactly like the one you wrote :-)

**Comment by asterite** [comment]
> Ah, highlighting the relevant name and then rightclicking that results in the above error. If I just rightclick the name without highlighting then it works.

Strange! I can't reproduce that. But it also didn't work for me the first time... for some reason I had two lsp servers runnings, so maybe it was using the old one that didn't have that functionality.

**Review by TomAFrench — APPROVED** [review]
LGTM

**Review by jfecher — APPROVED** [review]
